server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/GetByProjectAction.java

```diff
@@ -0,0 +1,118 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.qualitygate.ws;
+
+import java.util.Optional;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.db.qualitygate.QualityGateDto;
+import org.sonar.server.component.ComponentFinder;
+import org.sonar.server.component.ComponentFinder.ParamNames;
+import org.sonar.server.qualitygate.QualityGates;
+import org.sonarqube.ws.WsQualityGates.GetByProjectWsResponse;
+
+import static org.sonar.core.util.Uuids.UUID_EXAMPLE_01;
+import static org.sonar.server.ws.KeyExamples.KEY_PROJECT_EXAMPLE_001;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+import static org.sonarqube.ws.client.qualitygate.QualityGatesWsParameters.PARAM_PROJECT_ID;
+import static org.sonarqube.ws.client.qualitygate.QualityGatesWsParameters.PARAM_PROJECT_KEY;
+
+public class GetByProjectAction implements QualityGatesWsAction {
+  private final DbClient dbClient;
+  private final ComponentFinder componentFinder;
+  private final QualityGates qualityGates;
+
+  public GetByProjectAction(DbClient dbClient, ComponentFinder componentFinder, QualityGates qualityGates) {
+    this.dbClient = dbClient;
+    this.componentFinder = componentFinder;
+    this.qualityGates = qualityGates;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("get_by_project")
+      .setInternal(true)
+      .setSince("6.1")
+      .setDescription("Get the quality gate of a project.<br> " +
+        "Either project id or project key must be provided, not both.")
+      .setResponseExample(getClass().getResource("get_by_project-example.json"))
+      .setHandler(this);
+
+    action.createParam(PARAM_PROJECT_ID)
+      .setDescription("Project id")
+      .setExampleValue(UUID_EXAMPLE_01);
+
+    action.createParam(PARAM_PROJECT_KEY)
+      .setDescription("Project key")
+      .setExampleValue(KEY_PROJECT_EXAMPLE_001);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      ComponentDto project = componentFinder.getByUuidOrKey(dbSession, request.param(PARAM_PROJECT_ID), request.param(PARAM_PROJECT_KEY), ParamNames.PROJECT_ID_AND_KEY);
```

Comment:
check that user has project "Browse" or "Administer" permission or global "Administer" permission